packages/context/src/context-watcher.ts

```diff
@@ -0,0 +1,130 @@
+import {ResolutionSession} from './resolution-session';
+import {resolveList} from './value-promise';
+const debug = debugFactory('loopback:context:watcher');
+
+/**
+ * Watching a given context chain to maintain a live list of matching bindings
+ * and their resolved values within the context hierarchy.
+ *
+ * This class is the key utility to implement dynamic extensions for extension
+ * points. For example, the RestServer can react to `controller` bindings even
+ * they are added/removed/updated after the application starts.
+ *
+ */
+export class ContextWatcher<T = unknown> implements ContextListener {
+  protected _cachedBindings: Readonly<Binding<T>>[] | undefined;
+  protected _cachedValues: T[] | undefined;
+  private _subscription: Subscription | undefined;
+
+  constructor(
+    protected readonly ctx: Context,
+    public readonly filter: BindingFilter,
+  ) {}
+
+  /**
+   * Start watching events from the context
+   */
+  watch() {
+    debug('Start watching context %s', this.ctx.name);
+    return (this._subscription = this.ctx.subscribe(this));
+  }
+
+  /**
+   * Stop watching events from the context
+   */
+  unwatch() {
+    debug('Stop watching context %s', this.ctx.name);
+    if (this._subscription && !this._subscription.closed) {
+      this._subscription.unsubscribe();
+      this._subscription = undefined;
+    }
+  }
+
+  /**
+   * Get the list of matched bindings. If they are not cached, it tries to find
+   * them from the context.
+   */
+  get bindings(): Readonly<Binding<T>>[] {
+    debug('Reading bindings');
+    if (this._cachedBindings == null) {
+      this._cachedBindings = this.findBindings();
+    }
+    return this._cachedBindings;
+  }
+
+  /**
+   * Find matching bindings and refresh the cache
+   */
+  protected findBindings() {
+    debug('Finding matching bindings');
+    this._cachedBindings = this.ctx.find(this.filter);
+    return this._cachedBindings;
+  }
+
+  /**
+   * Listen on `bind` or `unbind` and invalidate the cache
+   */
+  listen(event: ContextEventType, binding: Readonly<Binding<unknown>>) {
+    this.reset();
+  }
+
+  /**
+   * Reset the watcher by invalidating its cache
+   */
+  reset() {
+    debug('Invalidating cache');
+    this._cachedBindings = undefined;
+    this._cachedValues = undefined;
+  }
+
+  /**
+   * Resolve values for the matching bindings
+   * @param session Resolution session
+   */
+  resolve(session?: ResolutionSession) {
+    debug('Resolving values');
+    return resolveList(this.bindings, b => {
+      return b.getValue(this.ctx, ResolutionSession.fork(session));
+    });
+  }
+
+  /**
+   * Get the list of resolved values. If they are not cached, it tries to find
+   * and resolve them.
+   */
+  async values(): Promise<T[]> {
+    debug('Reading values');
+    // [REVIEW] We need to get values in the next tick so that it can pick up
+    // binding changes as `Context` publishes such events in `process.nextTick`
+    return new Promise<T[]>(resolve => {
```

Comment:
Have you considered to use Promise microqueue instead of process.nextTick?

I think this should do the trick.

```ts
await Promise.resolve();
if (this._cachedValues == null) {
  this._cachedValues = await this.resolve();
}
return this._cachedValues();
```

If this does not work, then it would be still better to move the low-level complexity of converting nextTick callback into promises to a different helper function.

```ts
// usage
await nextTick();
if (this._cachedValues == null) {
  this._cachedValues = await this.resolve();
}
return this._cachedValues();

// impl - can we use util.promisify?
const nextTick = util.promisify(process.nextTick);

// if not
function nextTick() {
  return new Promise(resolve => process.nextTick(resolve));
}
```

Reply:
`await Promise.resolve();` does not work as `Promise.resolve()` will be resolved immediately.

Reply:
> `await Promise.resolve();` does not work as `Promise.resolve()` will be resolved immediately.

I am afraid that is not entirely correct. Even though the promise is resolved "immediately", any code consuming the resolved value will run in the next turn of the Promise micro-queue.

Consider the following example:

```js
Promise.resolve()
  .then(() => console.log('RESOLVED'));
console.log('ORIGINAL TICK');
```

On Node.js 10.15.0, I get the following output:

```
$ node x.js
ORIGINAL TICK
RESOLVED
```

This means that the code on the original stack (event loop tick) is run to completion before the promise is resolved. Another example using `await` instead:

```
async function later() {
  await Promise.resolve();
  console.log('LATER');
}

later();
console.log('NOW');
```

The output:

```
$ node x.js
NOW
LATER
```

Reply:
Context related events are emitted using `process.nextTick()`. Switching to `await Promise.resolve()` breaks the test cases. Maybe it has something to do with how next tick queue and promise queue are handled per https://jsblog.insiderattack.net/promises-next-ticks-and-immediates-nodejs-event-loop-part-3-9226cbe7a6aa

Reply:
I did some tracing and it turned out as follows:

1. Node.js processes nextTick queue, then promise queue, then nextTick again.
2. If we use `await Promise.resolve()`, the new bindings are not picked up as it's still in the loop of processing promise queue.
3. `await process.nextTick` gives new bindings a chance to be notified.

Reply:
Sounds good. Thank you for researching this topic, I learned something new here.